### PR TITLE
OCPBUGS-28370: Skip SG deletion if it has dependancy violation

### PR DESCRIPTION
**Pull request by Patryk-Stefanski**
**What this PR does / why we need it**:  Skips SG deletion to proceed with cluster deletion if there is a dependency violation with the SG (SG referenced by other non-hypershift controller resources), Also bubbles up an error to the HC to point to specific SG that hasn't been deleted.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes # [OCPBUGS-28370](https://issues.redhat.com/browse/OCPBUGS-28370)

HC CR remains on cluster with following error in status
```
2024-02-22T11:33:29Z'
      message: 'failed to delete AWS default security group due to dependency violation, security group: sg-0b9093871a'
      observedGeneration: 4
      reason: AWSError
      status: 'False'
      type: AWSDefaultSecurityGroupDeleted
```

CLI 
```
hypershift destroy cluster aws --name pstefans-hosted --aws-creds $AWS_CREDS --namespace clusters --aws-infra-grace-period 30s
2024-02-22T12:15:27Z    INFO    Found hosted cluster    {"namespace": "clusters", "name": "pstefans-hosted"}
2024-02-22T12:15:27Z    INFO    Marking cleanup of cloud resources for hosted cluster   {"namespace": "clusters", "name": "pstefans-hosted"}
2024-02-22T12:15:27Z    INFO    Updated hosted cluster  {"namespace": "clusters", "name": "pstefans-hosted"}
2024-02-22T12:15:27Z    INFO    Deleting hosted cluster {"namespace": "clusters", "name": "pstefans-hosted"}
2024-02-22T12:15:27Z    INFO    Destroying infrastructure       {"infraID": "***"}
2024-02-22T12:15:27Z    INFO    Infra destruction timeout set to 30 s
2024-02-22T12:15:35Z    INFO    WARNING: error during destroy, will retry       {"error": "[DependencyViolation: resource sg-0b9093871a75475ba has a dependent object\n\tstatus code: 400, request id: 37b78598-a15d-4917-af61-8a93f0f87d5c, DependencyViolation: The subnet 'subnet-0c0b90ae8487bf' has dependencies and cannot be deleted.\n\tstatus code: 400, request id: e1f7889a-1b4a-40cd-b452-c29f24ea2]"}
2024-02-22T12:15:47Z    INFO    WARNING: error during destroy, will retry       {"error": "[DependencyViolation: resource sg-0b9093871a754 has a dependent object\n\tstatus code: 400, request id: 2c728cf4-5063-4bd9-a1dc-fb19311f88ca, DependencyViolation: The subnet 'subnet-0c0b90ae8487b' has dependencies and cannot be deleted.\n\tstatus code: 400, request id: 555604a2-8bd1-46f0-b7e5-27a3b2ee61bf]"}
```
*Note*: if `--aws-infra-grace-period` isnt specified deletion wont timeout

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 1]
```diff
@@ -316,6 +316,9 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
 		if shouldCleanupCloudResources(r.Log, hostedControlPlane) {
 			if err := r.destroyAWSDefaultSecurityGroup(ctx, hostedControlPlane); err != nil {
 				condition.Message = "failed to delete AWS default security group"
+				if awsErrorCode(err) == "DependencyViolation" {
```
Comment (muraee) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:319]: maybe make the `destroyAWSDefaultSecurityGroup`  function return the aws error code, so we don't have to parse multiple times

Comment (Patryk-Stefanski) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:319]: that function returns multiple other formatted errors and I am not sure if the AWS errors  would be as specific.

Comment (muraee) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:319]: I meant in addition to the error

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 2]
```diff
@@ -4640,19 +4646,19 @@ func createAWSDefaultSecurityGroup(ctx context.Context, ec2Client ec2iface.EC2AP
 	return sgID, nil
 }
 
-func (r *HostedControlPlaneReconciler) destroyAWSDefaultSecurityGroup(ctx context.Context, hcp *hyperv1.HostedControlPlane) error {
+func (r *HostedControlPlaneReconciler) destroyAWSDefaultSecurityGroup(ctx context.Context, hcp *hyperv1.HostedControlPlane) (error, string) {
```
Comment (muraee) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:4649]: nit: usually the error is returned last

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 3]
```diff
@@ -324,8 +327,11 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
 					return ctrl.Result{}, fmt.Errorf("failed to update status on hcp for security group deletion: %w. Condition error message: %v", err, condition.Message)
 				}
 
-				if awsErrorCode(err) == "UnauthorizedOperation" {
+				if code == "UnauthorizedOperation" {
 					r.Log.Info("Skipping AWS default security group deletion because the operator is not authorized to delete it.")
+				}
+				if code == "DependencyViolation" {
+					r.Log.Info("Skipping AWS default security group deletion because of dependency violation.")
```
Comment (bryan-cox) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:334]: Should this be a warning instead of `Info`?

**Pull request merged**